zaza/openstack/charm_tests/ceph/tests.py

```diff
@@ -690,28 +702,162 @@ def expected_apps(self):
+        hostname = unit_hostnames[unit_name]
+        return 'radosgw-admin --id=rgw.{} '.format(hostname)
+
     @tenacity.retry(wait=tenacity.wait_exponential(multiplier=10, max=300),
                     reraise=True, stop=tenacity.stop_after_attempt(10),
                     retry=tenacity.retry_if_exception_type(AssertionError))
-    def wait_for_sync(self, application):
+    def wait_for_sync(self, application, isPrimary=False):
         """Wait for slave to secondary to show it is in sync."""
         juju_units = zaza_model.get_units(application)
         unit_hostnames = generic_utils.get_unit_hostnames(juju_units)
-        sync_states = []
-        sync_check_str = 'data is caught up with source'
+        sync_states = list()
+        data_check = 'data is caught up with source'
+        meta_primary = 'metadata sync no sync (zone is master)'
+        meta_secondary = 'metadata is caught up with master'
+        meta_check = meta_primary if isPrimary else meta_secondary
         for unit_name, hostname in unit_hostnames.items():
             key_name = "rgw.{}".format(hostname)
-            cmd = 'radosgw-admin --id={} zone sync status'.format(key_name)
+            cmd = 'radosgw-admin --id={} sync status'.format(key_name)
             stdout = zaza_model.run_on_unit(unit_name, cmd).get('Stdout', '')
-            sync_states.append(sync_check_str in stdout)
+            sync_states.append((data_check in stdout) and
+                               (meta_check in stdout))
         assert all(sync_states)
 
-    def test_processes(self):
+    @tenacity.retry(wait=tenacity.wait_exponential(multiplier=1, max=60),
+                    reraise=True, stop=tenacity.stop_after_attempt(12))
+    def fetch_rgw_object(self, target_client, container_name, object_name):
+        '''
+        Fetch RGW object content
+        :param target_client: boto3 client object configured for an endpoint.
+        :param container_name: RGW bucket name for desired object.
+        :param object_name: Object name for desired object.
+        '''
+        return target_client.Object(container_name, object_name).get()[
+            'Body'
+        ].read().decode('UTF-8')
+
+    def promote_rgw_to_primary(self, app_name:str):
+        '''Promote provided app to Primary and update period at new secondary'''
+        if app_name is self.primary_rgw_app:
+            new_secondary = self.secondary_rgw_unit
+        else:
+            new_secondary = self.primary_rgw_unit
+
+        # Promote to Primary
+        zaza_model.run_action_on_leader(
+            app_name,
+            'promote',
+            action_params={},
+        )
+
+        # Period Update Commit new secondary.
+        cmd = self.get_rgwadmin_cmd_skeleton(new_secondary)
+        zaza_model.run_on_unit(
+            new_secondary, cmd + 'period update --commit'
+        ).get('Stdout', '')
+
+    def get_client_keys(self, rgw_app_name=None):
```

Comment:
This seems like it may be more generally useful; however, we haven't needed it yet, so maybe leave it here and we can move it later

Reply:
ACK